rclcpp/test/rclcpp/executors/test_executors_callback_group_behavior.cpp

```diff
@@ -0,0 +1,140 @@
+// Copyright 2024 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * This test checks that when callback groups go out of scope, that their associated executable
+ * entities should not be returned as valid executables.
+ *
+ * The test makes use of a bit of executor internals, but is meant to prevent regressions of behavior.
+ * Ref: https://github.com/ros2/rclcpp/issues/2474
+ */
+
+#include <gtest/gtest.h>
+
+#include <chrono>
+#include <rclcpp/callback_group.hpp>
+#include <rclcpp/executor.hpp>
+#include <rclcpp/node.hpp>
+
+class CustomExecutor: public rclcpp::Executor
+{
+public:
+  explicit CustomExecutor(const rclcpp::ExecutorOptions & options = rclcpp::ExecutorOptions())
+    : rclcpp::Executor(options)
+  {
+  }
+
+  ~CustomExecutor() override = default;
+
+  void spin() override {};
+
+  void collect() {
+    this->collect_entities();
+  }
+
+  void wait() {
+    this->wait_for_work(std::chrono::milliseconds(10));
+  }
+
+  size_t collected_timers() const {
+    return this->current_collection_.timers.size();
+  }
+
+  rclcpp::AnyExecutable next() {
+    rclcpp::AnyExecutable ret;
+    this->get_next_ready_executable(ret);
+    return ret;
+  }
+};
+
+
+TEST(TestCallbackGroup, ValidCbg)
```

Comment:
nitpick, we typically use snakecase for the test name and I'd appreciate it if we try to use the full name for things (helps with translation for people who don't speak english and pattern matching), e.g.:

```suggestion
TEST(TestCallbackGroup, valid_callback_group)
```

Reply:
According to gtest docs, `PascalCase` should be used

> TEST() arguments go from general to specific. The first argument is the name of the test suite, and the second argument is the test’s name within the test suite. Both names must be valid C++ identifiers, and they should not contain any underscores (_)

from https://google.github.io/googletest/primer.html#simple-tests

Reply:
That's interesting, I never knew that, looking into it, it seems like it doesn't matter now, but could in the future: https://stackoverflow.com/a/34543433

I think we should refactor them all to avoid this condition, but obviously not in this pr. @clalancette

Reply:
Wow, I had no idea.  We have literally thousands of gtests in ROS 2 using underscore.

That said, I actually think we should stick with our current snake-case style, even though it is slightly wrong.  To me it seems better to be consistent, and if we want to fix this (pretty theoretical) issue, we can do it all at once.